homeassistant/components/thermopro/button.py

```diff
@@ -0,0 +1,169 @@
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.util.dt import now
+
+from .const import DOMAIN, SIGNAL_AVAILABILITY_UPDATED, SIGNAL_DATA_UPDATED
+
+_LOGGER = logging.getLogger(__name__)
+
+DATETIME_UPDATE = ButtonEntityDescription(
+    key="datetime",
+    translation_key="set_datetime",
+    icon="mdi:calendar-clock",
+    entity_category=EntityCategory.CONFIG,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up the demo button platform."""
+    assert entry.unique_id is not None
+
+    set_datetime_entity = None
+
+    @callback
+    def _async_on_data_updated(
+        data: ThermoProBluetoothDeviceData,
+        service_info: BluetoothServiceInfoBleak,
+        update: SensorUpdate,
+    ) -> None:
+        nonlocal set_datetime_entity
+        _LOGGER.debug(
+            "got update data=%s update=%s service_info=%s", data, update, service_info
+        )
+
+        assert entry.unique_id is not None
+        assert None in update.devices
+        assert update.devices[None].model is not None
+
+        if update.devices[None].model not in ("TP358", "TP393"):
+            return
+
+        if not set_datetime_entity:
+            name = update.devices[None].name
+            assert name is not None
+            set_datetime_entity = ThermoProDateTimeButtonEntity(
+                entry_id=entry.entry_id,
+                address=entry.unique_id,
+                device_name=name,
+                description=DATETIME_UPDATE,
+            )
+
+            async_add_entities(
+                [
+                    set_datetime_entity,
+                ]
+            )
+
+        if service_info.connectable and not set_datetime_entity.available:
+            _LOGGER.debug("sending availability '%s' for %s", True, entry.entry_id)
+            async_dispatcher_send(
+                hass,
+                f"{SIGNAL_AVAILABILITY_UPDATED}_{entry.entry_id}_button_datetime",
+                True,
+            )
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, f"{SIGNAL_DATA_UPDATED}_{entry.entry_id}", _async_on_data_updated
+        )
+    )
+
+    @callback
+    def _async_on_unavailable(service_info: BluetoothServiceInfoBleak) -> None:
+        _LOGGER.debug("service info unavailable %s", service_info)
+        if set_datetime_entity:
+            _LOGGER.debug("sending availability '%s' for %s", False, entry.entry_id)
+            async_dispatcher_send(
+                hass,
+                f"{SIGNAL_AVAILABILITY_UPDATED}_{entry.entry_id}_button_datetime",
+                False,
+            )
+
+    entry.async_on_unload(
+        async_track_unavailable(
+            hass, _async_on_unavailable, entry.unique_id, connectable=True
+        )
+    )
```

Comment:
Thought about this one some more, and we can simplify it a bit since we don't need this anywhere else we could setup the in `async_added_to_hass` instead and use `self.async_on_unload(async_track_unavailable(self.hass, self._async_on_unavailable, self.address, connectable=True)` and implement `_async_on_unavailable` on the entity.

Reply:
Wouldn't we need the dispatcher stuff anyways to note the re-availability of the entity? I implemented it now in this way, directly using the listener inside of the entity for unavailability and the dispatcher for availability.

Can you give me a hint about the reavailability test? Is it needed? I can't get the injection of advertisments to work after the time manipulation - probably due to that somehow messing with the scanner in a way I don't get.

Reply:
Once I cleaned up things, your test seems to work